wpimath/src/main/java/edu/wpi/first/math/estimator/PoseEstimator3d.java

```diff
@@ -0,0 +1,616 @@
+   * automatically takes care of offsetting the gyro angle.
+   *
+   * @param gyroAngle The angle reported by the gyroscope.
+   * @param wheelPositions The current encoder readings.
+   * @param poseMeters The position on the field that your robot is at.
+   */
+  public void resetPosition(Rotation3d gyroAngle, T wheelPositions, Pose3d poseMeters) {
+    // Reset state estimate and error covariance
+    m_odometry.resetPosition(gyroAngle, wheelPositions, poseMeters);
+    m_odometryPoseBuffer.clear();
+    m_visionUpdates.clear();
+    m_poseEstimate = m_odometry.getPose3dMeters();
+  }
+
+  /**
+   * Resets the robot's pose.
+   *
+   * @param pose The pose to reset to.
+   */
+  public void resetPose(Pose2d pose) {
+    resetPose(new Pose3d(pose));
+  }
+
+  /**
+   * Resets the robot's pose.
+   *
+   * @param pose The pose to reset to.
+   */
+  public void resetPose(Pose3d pose) {
+    m_odometry.resetPose(pose);
+    m_odometryPoseBuffer.clear();
+    m_visionUpdates.clear();
+    m_poseEstimate = m_odometry.getPose3dMeters();
+  }
+
+  /**
+   * Resets the robot's translation.
+   *
+   * @param translation The pose to translation to.
+   */
+  public void resetTranslation(Translation2d translation) {
+    resetTranslation(new Translation3d(translation.getX(), translation.getY(), 0));
+  }
+
+  /**
+   * Resets the robot's translation.
+   *
+   * @param translation The pose to translation to.
+   */
+  public void resetTranslation(Translation3d translation) {
+    m_odometry.resetTranslation(translation);
+    m_odometryPoseBuffer.clear();
+    m_visionUpdates.clear();
+    m_poseEstimate = m_odometry.getPose3dMeters();
+  }
+
+  /**
+   * Resets the robot's rotation.
+   *
+   * @param rotation The rotation to reset to.
+   */
+  public void resetRotation(Rotation2d rotation) {
+    resetRotation(new Rotation3d(0, 0, rotation.getRadians()));
+  }
+
+  /**
+   * Resets the robot's rotation.
+   *
+   * @param rotation The rotation to reset to.
+   */
+  public void resetRotation(Rotation3d rotation) {
+    m_odometry.resetRotation(rotation);
+    m_odometryPoseBuffer.clear();
+    m_visionUpdates.clear();
+    m_poseEstimate = m_odometry.getPose3dMeters();
+  }
+
+  /**
+   * Gets the estimated robot pose.
+   *
+   * @return The estimated robot pose in meters.
+   */
+  public Pose2d getEstimatedPosition() {
+    return m_poseEstimate.toPose2d();
+  }
+
+  /**
+   * Gets the estimated robot pose.
+   *
+   * @return The estimated robot pose in meters.
+   */
+  public Pose3d getEstimatedPosition3d() {
+    return m_poseEstimate;
+  }
+
+  /**
+   * Return the pose at a given timestamp, if the buffer is not empty.
+   *
+   * @param timestampSeconds The pose's timestamp in seconds.
+   * @return The pose at the given timestamp (or Optional.empty() if the buffer is empty).
+   */
+  public Optional<Pose2d> sampleAt(double timestampSeconds) {
```

Comment:
We could just return `Optional<Pose3d>` from this function, remove the `sampleAt3d()` overload, and let the user call `toPose2d()` on the unwrapped value if they need to.